src/middlewared/middlewared/plugins/sysdataset.py

```diff
@@ -244,110 +316,142 @@ async def destination_pool_error(self, new_pool):
 
-        acltype = await self.middleware.call('zfs.dataset.query', [('id', '=', config['basename'])])
-        if acltype and acltype[0]['properties']['acltype']['value'] != 'off':
-            await self.middleware.call(
+        ds_mntinfo = filter_list(mntinfo, [['mount_source', '=', config['basename']]])
+        if ds_mntinfo:
+            acl_enabled = 'POSIXACL' in ds_mntinfo[0]['super_opts'] or 'NFSV4ACL' in ds_mntinfo[0]['super_opts']
+        else:
+            ds = self.middleware.call_sync('zfs.dataset.query', [('id', '=', config['basename'])])
+            acl_enabled = ds and ds[0]['properties']['acltype']['value'] != 'off'
+
+        if acl_enabled:
+            self.middleware.call_sync(
                 'zfs.dataset.update', config['basename'], {'properties': {'acltype': {'value': 'off'}}}
             )
 
         if mounted is None:
-            mounted = await self.__mount(config['pool'], config['uuid'])
+            mounted = self.__mount(config['pool'], config['uuid'])
 
         corepath = f'{SYSDATASET_PATH}/cores'
         if os.path.exists(corepath):
             os.chmod(corepath, 0o775)
 
-            if await self.middleware.call('keyvalue.get', 'run_migration', False):
+            if self.middleware.call_sync('keyvalue.get', 'run_migration', False):
                 try:
                     cores = Path(corepath)
                     for corefile in cores.iterdir():
                         corefile.unlink()
                 except Exception:
                     self.logger.warning("Failed to clear old core files.", exc_info=True)
 
-            await run('umount', '/var/lib/systemd/coredump', check=False)
+            subprocess.run(['umount', '/var/lib/systemd/coredump'], check=False)
             os.makedirs('/var/lib/systemd/coredump', exist_ok=True)
-            await run('mount', '--bind', corepath, '/var/lib/systemd/coredump')
+            subprocess.run(['mount', '--bind', corepath, '/var/lib/systemd/coredump'])
 
-        await self.middleware.call('etc.generate', 'glusterd')
+        self.middleware.call_sync('etc.generate', 'glusterd')
 
         if mounted:
-            await self.middleware.call('smb.setup_directories')
+            self.middleware.call_sync('systemdataset._post_setup_service_restart')
 
-            # There is no need to wait this to finish
-            # Restarting rrdcached will ensure that we start/restart collectd as well
-            asyncio.ensure_future(self.middleware.call('service.restart', 'rrdcached'))
-            asyncio.ensure_future(self.middleware.call('service.restart', 'syslogd'))
+        return self.middleware.call_sync('systemdataset.config')
 
-            # The following should be backgrounded since they may be quite
-            # long-running.
-            await self.middleware.call('smb.configure', False)
+    @private
+    def query_pool_for_system_dataset(self, exclude_pool):
+        for p in self.middleware.call_sync('pool.query'):
```

Comment:
zfs.pool.query_imported_fast instead of pool.query?

Reply:
Sure. I also now have set `retrieve_children: False` which should also help speed things up.